src/dmd/target.d

```diff
@@ -1523,19 +1532,51 @@ struct Triple
         else
         {
             unknown(_os, "operating system");
-            return Target.OS.Freestanding;
+            return Target.OS.none;
         }
-        while (_os.length)
+
+        bool overflow;
+        auto major = parseNumber(_os, overflow);
+        if (overflow || major >= 255)
         {
-            if (!('0' < _os[0] && _os[0] < '9'))
-                break;
-            osMajor *= 10;
-            osMajor = cast(ubyte)((_os[0] - '0') + osMajor);
-            _os = _os[1 .. $];
+            error(Loc.initial, "OS version overflowed max of 254");
+            major = 255;
         }
+        osMajor = cast(ubyte)major;
+
+        /* Note that anything after the number up to the end or '-',
+         * such as '.3.4.hello.betty', is ignored
+         */
+
         return os;
     }
 
+    /*******************************
+     * Parses a decimal number out of the str and returns it.
+     * Params:
+     *  str = string to parse the number from, updated to text after the number
+     *  overflow = set to true iff an overflow happens
+     * Returns:
+     *  parsed number
+     */
+    private pure static
+    uint parseNumber(ref const(char)[] str, ref bool overflow)
```

Comment:
On second thought, this is just silly to introduce given that we already have [dmd.utils.parseDigits](https://github.com/dlang/dmd/blob/8e6e7ff260f18a73d835abd3011a0b279c062a25/src/dmd/utils.d#L227).

Reply:
On third thought, parseDigits fails if it encounters a non-digit, while parseNumber simply stops. This matters if the suffix is significant (going back to my comment about ignoring parts of the triples and having a fuzzy specification).

Reply:
Indeed, but that behaviour can be dealt with either by an extra parameer to `parseDigits` to allow flipping between behaviours, or lookahead and pass `parseDigits` a slice that only contains digits.

Reply:
That is a fine idea for a subsequent refactoring PR.